src/main/java/uk/ac/cam/cl/dtg/isaac/api/managers/QuizQuestionManager.java

```diff
@@ -81,4 +85,56 @@ public void recordQuestionAttempt(QuizAttemptDTO quizAttempt, QuestionValidation
 
         this.quizQuestionAttemptManager.registerQuestionAttempt(quizAttempt.getId(), questionResponseDO);
     }
+
+    /**
+     * This method will ensure any user question attempt information available is used to augment this question object.
+     *
+     * It will also ensure that any personalisation of questions is affected (e.g. randomised multichoice elements).
+     *
+     * Note: It will not do anything to related content
+     *  @param quiz
+     *            - to augment - this object may be mutated as a result of this method. i.e BestAttempt field set on
+     *            question DTOs.
+     * @param quizAttempt
+     *            - which attempt at the quiz to get attempts for.
+     * @param user
+     *            - which user to augment the questions for.
+     * @param includeCorrect
+     *            - include whether the answers are correct.
+     */
+    public void augmentQuestionsForUser(IsaacQuizDTO quiz, QuizAttemptDTO quizAttempt, RegisteredUserDTO user, boolean includeCorrect) throws SegueDatabaseException {
+        List<QuestionDTO> questionsToAugment = QuestionManager.extractQuestionObjects(quiz);
+
+        this.augmentQuestionObjectWithAttemptInformation(quizAttempt, questionsToAugment, includeCorrect);
+
+        questionManager.shuffleChoiceQuestionsChoices(user.getId().toString(), questionsToAugment);
+    }
+
+    /**
+     * Modify the questions in a quiz such that it contains bestAttempt information if we can provide it.
+     *
+     * When we say bestAttempt, we actually mean latest attempt.
+     *  @param quizAttempt
+     *            - which attempt at the quiz to get attempts for.
+     * @param questionsToAugment
+     *            - list of question objects to modify.
+     * @param includeCorrect
+     *            - include whether the answers are correct.
+     */
+    private void augmentQuestionObjectWithAttemptInformation(QuizAttemptDTO quizAttempt, List<QuestionDTO> questionsToAugment, boolean includeCorrect) throws SegueDatabaseException {
+        Map<String, List<QuestionValidationResponse>> answers = quizQuestionAttemptManager.getAnswers(quizAttempt.getId());
+
+        for (QuestionDTO question : questionsToAugment) {
+            List<QuestionValidationResponse> questionAttempts = answers.get(question.getId());
+
+            // The latest answer is the only answer we consider.
+            if (questionAttempts.size() > 0) {
+                QuestionValidationResponseDTO lastAttempt = questionManager.convertQuestionValidationResponseToDTO(questionAttempts.get(questionAttempts.size() - 1));
+                if (!includeCorrect) {
+                    lastAttempt.setCorrect(null);
```

Comment:
Will this not leave the `explanation` field from the validation response still present, which might explain the correct answer or inform them of the correct/incorrectness. What about `QuantityValidationResponseDTO` s which have `correctValue` and `correctUnits`? (There are other subclasses, but that is the only one that is exposed to the user other than the base class by `QuestionValidationResponseOrikaConverter`).

I wonder whether the safest thing to do is to manually create a new `QuestionValidationResponseDTO`, with only the `questionId`, `answer` and `dateAttempted` fields copied over from the actual DO, using a new private method?

Reply:
Yes, I think I added `lastAttempt.setExplanation(null);` but maybe that was on another PR?

In any case, I hadn't thought about subclasses. I think manually copying out of the DO into a DTO with those fields would be safer, yes. I'll do that.